tracetools_trace/tracetools_trace/tools/lttng_impl.py

```diff
@@ -55,6 +58,48 @@ def is_kernel_tracer_available() -> bool:
     return not isinstance(lttngpy.get_tracepoints(domain_type=lttngpy.LTTNG_DOMAIN_KERNEL), int)
 
 
+def is_kernel_paranoid_perf_event(context_fields: Set[str]) -> Tuple[bool, Optional[str]]:
+    """
+    Check if the kernel is paranoid about perf events, if any.
+
+    The current kernel may not allow perf events (exposed by LTTng as context fields) to be used
+    depending on the perf_event_paranoid value (/proc/sys/kernel/perf_event_paranoid), see:
+    https://www.kernel.org/doc/Documentation/sysctl/kernel.txt. In that case, the resulting trace
+    will be unexpectedly empty.
+
+    Note: the CAP_SYS_ADMIN capability may still allow perf events to be used even if the kernel is
+    paranoid. However, we are not checking for that here.
+
+    :param: context_fields: the names of requested context fields
+    :return: a (is_paranoid, message) tuple, where:
+        is_paranoid: `True` if kernel is paranoid about given context fields, otherwise `False`
+        message: a message explaining the paranoia (only applicable if `is_paranoid` is `True`)
+    """
+    print('is_kernel_paranoid_perf_event context_fields=', context_fields)
+    # For now, just check for 'perf:thread:*' perf events
+    has_perf_thread_context_fields = any(
+        # Wildcards (*) are not supported for context fields
+        context_field.startswith('perf:thread:')
+        for context_field in context_fields
+    )
+    if not has_perf_thread_context_fields:
+        return False, None
+    path_perf_event_paranoid = '/proc/sys/kernel/perf_event_paranoid'
+    paranoid_value_str = Path(path_perf_event_paranoid).read_text().strip()
```

Comment:
i would add the try&catch statement here just in case? even with linux system, this depends on the kernel configuration i guess... i would catch `FileNotFoundError`, `PermissionError`.

Reply:
what i only think of is that user intentionally disable instrument flag for kernel config. which is the case that user know instruments cannot be used, so maybe no need to catch the exception, i guess. maybe you can ignore this comment.

Reply:
That's a good point, but then it becomes harder to detect. I think we can proceed with this, which should detect most cases, and improve it later if we think it's needed.